### PR TITLE
chore: Tracking corrections, refactor address verification flow

### PR DESCRIPTION
**Pull request by erikdstock**
The type of this PR is: **chore**
Resolves [EM-1402]

cc @artsy/emerald-devs 
### Description

This PR updates our address verification flow with 🔒[corrected tracking values](https://artsy.slack.com/archives/C02JHHHKP5K/p1693242656034689) and refactors to remove unnecessary `useState/useEffect` hooks, which may be contributing to some infinite loops we've seen in testing. 

This PR replaces #12736 which grew too big, so I am reopening with a clean slate.

### CODE REVIEW

File: src/Apps/Order/Components/AddressVerificationFlow.tsx [hunk 1]
```diff
@@ -195,90 +113,129 @@ const AddressVerificationFlow: React.FC<AddressVerificationFlowProps> = ({
...
+      onChosenAddress(AddressVerifiedBy.ARTSY, addressOptions[0].address!)
+      didLoad.current = true
+    } else {
+      setShowModal(true)
+      didLoad.current = true
+
+      trackViewedModal({
+        subject: modalTitle!,
+      })
+    }
```
Comment (erikdstock) [inline src/Apps/Order/Components/AddressVerificationFlow.tsx:173]: a previous large useEffect that was running unpredictably has been mostly moved into the imperative outer layer of the code. Instead, we put only the actual side effects (which should only occur once) in here, and make sure it only loads once.

Maybe the `didLoad.current = true` line should just move below L172.

File: src/Apps/Order/Components/AddressVerificationFlow.tsx [hunk 2]
```diff
@@ -479,3 +439,73 @@ export const AddressVerificationFlowQueryRenderer: React.FC<{
...
+          context_module: ContextModule.ordersShipping,
+          context_page_owner_type: OwnerType.ordersShipping,
+          context_page_owner_id: contextPageOwnerId,
+          user_id: userId,
+          flow: "user adding shipping address",
+          subject,
+        } as ValidationAddressViewed)
+      },
+      [contextPageOwnerId, trackEvent, userId]
+    ),
```
Comment (erikdstock) [inline src/Apps/Order/Components/AddressVerificationFlow.tsx:485]: tracking logic moved into this local helper to clean up the main component. I'm not sure if all these `useCallback()`s are necessary.

File: src/Apps/Order/Components/AddressVerificationFlow.tsx [hunk 3]
```diff
@@ -195,90 +113,129 @@ const AddressVerificationFlow: React.FC<AddressVerificationFlowProps> = ({
...
+    subject: string
+    option?: string
+  }) => {
+    trackClickedModal({
+      subject,
+      option,
+      label,
+    })
+    onClose()
+  }
```
Comment (erikdstock) [inline src/Apps/Order/Components/AddressVerificationFlow.tsx:217]: Realized that closing the modal is a different tracking event from choosing an address so i need to disambiguate this bit.

File: src/Apps/Order/Routes/Shipping/index.tsx [hunk 4]
```diff
@@ -757,208 +758,215 @@ export const ShippingRoute: FC<ShippingProps> = props => {
...
+                    <BorderedRadio value="SHIP" label="Shipping" />
+                    <BorderedRadio
+                      value="PICKUP"
+                      label="Arrange for pickup (free)"
+                      data-test="pickupOption"
+                    >
+                      <Collapse open={shippingOption === "PICKUP"}>
+                        <Text variant="xs" color="black60">
+                          After your order is confirmed, a specialist will
+                          contact you to coordinate pickup.
```
Comment (erikdstock) [inline src/Apps/Order/Routes/Shipping/index.tsx:799]: 🤔

File: src/Apps/Order/Routes/Shipping/index.tsx [hunk 5]
```diff
@@ -757,208 +758,215 @@ export const ShippingRoute: FC<ShippingProps> = props => {
...
-            flexDirection="column"
-            style={isCommittingMutation ? { pointerEvents: "none" } : {}}
-          >
-            {/* TODO: Make RadioGroup generic for the allowed values,
+    <AnalyticsContext.Provider
+      value={{
+        contextPageOwnerId: order.internalID,
+        contextPageOwnerType: OwnerType.ordersShipping,
+      }}
+    >
```
Comment (erikdstock) [inline src/Apps/Order/Routes/Shipping/index.tsx:766]: Need to double check that this is still necessary after #12776 which just merged.

**Review by erikdstock — COMMENTED** [review]
Small comments, I'll push followups shortly.

File: src/Apps/Order/Components/AddressVerificationFlow.tsx [hunk 6]
```diff
@@ -195,90 +113,129 @@ const AddressVerificationFlow: React.FC<AddressVerificationFlowProps> = ({
...
     }
+  }
+
+  const modalTitle = modalTitles[verificationPath]
+
+  const [selectedAddressKey, setSelectedAddressKey] = useState<
+    AddressOptionKey | undefined
+  >(addressOptions[0]?.key)
+
+  // perform only once on mount
```
Comment (damassi) [inline src/Apps/Order/Components/AddressVerificationFlow.tsx:155]: This comment isn't actually accurate, since there are dependencies in the `useEffect` array. If you want it to only perform once, need to explicitly use `[]`

File: src/Apps/Order/Components/AddressVerificationFlow.tsx [hunk 7]
```diff
@@ -479,3 +439,73 @@ export const AddressVerificationFlowQueryRenderer: React.FC<{
...
+        trackEvent({
+          action: ActionType.clickedValidationAddressOptions,
+          context_module: ContextModule.ordersShipping,
+          context_page_owner_type: OwnerType.ordersShipping,
+          context_page_owner_id: contextPageOwnerId,
+          user_id: userId,
+          subject,
+          ...(typeof subject !== "undefined" && { subject }),
+          ...(typeof option !== "undefined" && { option }),
+          ...(typeof label !== "undefined" && { label }),
```
Comment (damassi) [inline src/Apps/Order/Components/AddressVerificationFlow.tsx:505]: Can this be accomplished in a less ambiguous way?

File: src/Apps/Order/Routes/Shipping/index.tsx (hunk 5 above)
Comment (damassi) [reply src/Apps/Order/Routes/Shipping/index.tsx:766]: This is no longer valid as of that PR, as we've changed the API

File: src/Apps/Order/Components/AddressVerificationFlow.tsx (hunk 6 above)
Comment (erikdstock) [reply src/Apps/Order/Components/AddressVerificationFlow.tsx:155]: Ah, i guess in this case i don't need the ref - i thought there was a risk of double-firing events without it and was hesitant to disable the exhaustive deps rule but i can do it if you say its ok.

Comment (damassi) [reply src/Apps/Order/Components/AddressVerificationFlow.tsx:155]: If you want something to fire once, the effect fires on mount -- and if there are no deps to track, it will never fire again. The linter doesn't know your intent, so it will always warn.

File: src/Apps/Order/Components/AddressVerificationFlow.tsx (hunk 1 above)
Comment (erikdstock) [reply src/Apps/Order/Components/AddressVerificationFlow.tsx:173]: Moved down, then removed completely because there was nothing to wait for before loading - firing once on mount (`[]` deps array) does the trick.

File: src/Apps/Order/Routes/Shipping/index.tsx (hunk 5 above)
Comment (erikdstock) [reply src/Apps/Order/Routes/Shipping/index.tsx:766]: Thanks. I think I updated for that - Let me know if i missed something

Comment (damassi) [reply src/Apps/Order/Routes/Shipping/index.tsx]: Looks good!

File: src/Apps/Order/Components/AddressVerificationFlow.tsx [hunk 8]
```diff
@@ -479,3 +426,90 @@ export const AddressVerificationFlowQueryRenderer: React.FC<{
...
+    ),
+
+    trackClosedModal: useCallback(
+      ({ label, subject }: { label: string; subject: string }) => {
+        const event: ClickedCloseValidationAddressModal = {
+          action: ActionType.clickedCloseValidationAddressModal,
+          context_module: ContextModule.ordersShipping,
+          context_page_owner_type: OwnerType.ordersShipping,
+          context_page_owner_id: contextPageOwnerId!,
+          option: "close",
```
Comment (starsirius) [inline src/Apps/Order/Components/AddressVerificationFlow.tsx:506]: "close" is not a valid option from the [spreadsheet](https://docs.google.com/spreadsheets/d/1c65vugkF2C4YuXwBXno8aqsBh4B2fTwxiLmgY_MhjWc/edit#gid=0). Looks like this needs to be dynamic.

Screen Shot 2023-08-25 at 1 46 12 PM

File: src/Apps/Order/Components/AddressVerificationFlow.tsx [hunk 9]
```diff
@@ -195,90 +114,114 @@ const AddressVerificationFlow: React.FC<AddressVerificationFlowProps> = ({
...
+  return !showModal ? null : (
     <ModalDialog
       width={["100%", 590]}
       height={["100vh", "auto"]}
       m={[0, 2]}
-      title={modalTitles[modalType]}
-      onClose={handleClose}
+      title={modalTitle!}
+      onClose={() => {
+        handleCloseModal({ label: "close modal", subject: modalTitle! })
```
Comment (starsirius) [inline src/Apps/Order/Components/AddressVerificationFlow.tsx:221]: If I remember correctly, the label should be `null` when the modal is closed via clicking "X", clicking away, or hitting <ESC>. "close modal" is not a valid label from the [spreadsheet](https://docs.google.com/spreadsheets/d/1c65vugkF2C4YuXwBXno8aqsBh4B2fTwxiLmgY_MhjWc/edit#gid=0).

File: src/Apps/Order/Components/AddressVerificationFlow.tsx [hunk 10]
```diff
@@ -317,7 +260,12 @@ const AddressVerificationFlow: React.FC<AddressVerificationFlowProps> = ({
           <Spacer y={4} />
           <Flex width="100%" justifyContent="space-between">
             <Button
-              onClick={handleBackToEdit}
+              onClick={() => {
+                handleCloseModal({
```
Comment (starsirius) [inline src/Apps/Order/Components/AddressVerificationFlow.tsx:264]: If I understand correctly from the [spreadsheet](https://docs.google.com/spreadsheets/d/1c65vugkF2C4YuXwBXno8aqsBh4B2fTwxiLmgY_MhjWc/edit#gid=0), we want to track both the `ClickedValidationAddressOptions` and `ClickedCloseValidationAddressModal` event here.

File: src/Apps/Order/Components/AddressVerificationFlow.tsx [hunk 11]
```diff
@@ -328,13 +276,8 @@ const AddressVerificationFlow: React.FC<AddressVerificationFlowProps> = ({
...
               onClick={() => {
                 if (selectedAddressKey) {
-                  trackEvent({
-                    action_type: "clickedValidationAddress",
-                    context_module: ContextModule.ordersShipping,
-                    context_page_owner_type: OwnerType.ordersShipping,
-                    context_page_owner_id: contextPageOwnerSlug,
-                    user_id: userId,
-                    subject: "Confirm your delivery address",
+                  trackSelectedAddress({
```
Comment (starsirius) [inline src/Apps/Order/Components/AddressVerificationFlow.tsx:279]: Same as https://github.com/artsy/force/pull/12778/files?diff=unified&w=0#r1305979017.

File: src/Apps/Order/Components/AddressVerificationFlow.tsx [hunk 12]
```diff
@@ -372,13 +315,9 @@ const AddressVerificationFlow: React.FC<AddressVerificationFlowProps> = ({
...
             <Button
               onClick={() => {
-                trackEvent({
-                  action_type: "clickedValidationAddress",
-                  context_module: ContextModule.ordersShipping,
-                  context_page_owner_type: OwnerType.ordersShipping,
-                  context_page_owner_id: contextPageOwnerSlug,
-                  user_id: userId,
-                  subject: "Check your delivery address",
+                trackSelectedAddress({
```
Comment (starsirius) [inline src/Apps/Order/Components/AddressVerificationFlow.tsx:318]: Same as https://github.com/artsy/force/pull/12778/files?diff=unified&w=0#r1305979017.

File: src/Apps/Order/Components/AddressVerificationFlow.tsx [hunk 13]
```diff
@@ -389,7 +328,15 @@ const AddressVerificationFlow: React.FC<AddressVerificationFlowProps> = ({
               Use This Address
             </Button>
             <Spacer x={1} />
-            <Button onClick={handleEditAddress} flex={1}>
+            <Button
+              onClick={() => {
+                handleCloseModal({
```
Comment (starsirius) [inline src/Apps/Order/Components/AddressVerificationFlow.tsx:333]: Same as https://github.com/artsy/force/pull/12778/files?diff=unified&w=0#r1305979017.

File: src/Apps/Order/Components/AddressVerificationFlow.tsx [hunk 14]
```diff
@@ -195,90 +114,114 @@ const AddressVerificationFlow: React.FC<AddressVerificationFlowProps> = ({
...
+  useEffect(() => {
+    if (verificationPath === VerificationPath.ERROR_IMMEDIATE_CONFIRM) {
+      const fallbackOption = fallbackFromFormValues(verificationInput)
+      onChosenAddress(AddressVerifiedBy.USER, fallbackOption.address)
+    } else if (verificationPath === VerificationPath.IMMEDIATE_CONFIRM) {
+      onChosenAddress(AddressVerifiedBy.ARTSY, addressOptions[0].address!)
+    } else {
+      setShowModal(true)
+
+      trackViewedModal({
```
Comment (starsirius) [inline src/Apps/Order/Components/AddressVerificationFlow.tsx:166]: Question: firing the viewed event right after setting a state seems to be a few steps early than the modal is rendered to the user. Is this generally safe?

File: src/Apps/Order/Components/AddressVerificationFlow.tsx (hunk 9 above)
Comment (erikdstock) [reply src/Apps/Order/Components/AddressVerificationFlow.tsx:221]: Ah, i was looking into the palette docs and couldn't find a handler for clicking outside the modal so i decided to just put this for now. I also ran into an issue with some of these (like `option` where the cohesion scheme didn't permit null even though it was the specified value.

File: src/Apps/Order/Routes/__tests__/Shipping.jest.tsx [hunk 15]
```diff
@@ -206,11 +206,7 @@ class ShippingTestPage extends OrderAppTestPage {
   }
 }
 
-describe("Shipping [testing-library]", () => {
-  describe("US address verification", () => {})
-})
-
-describe("Shipping [enzyme]", () => {
+describe("Shipping", () => {
```
Comment (erikdstock) [inline src/Apps/Order/Routes/__tests__/Shipping.jest.tsx:209]: Removed these empty tests that i accidentally merged recently 🙊

**Review by erikdstock — COMMENTED** [review]
I met with Dayane to get final clarification on tracking behaviors we want here. In short:
- Track dismissing modal as `clickedCloseValidationAddressModal` with option as `null` and label as the 'back to edit/edit address' button text, or when closing via `x` as `null`. This label value as `null` violates the cohesion schema and will be ticketed as a followup, using coercions for now.
- Track choosing an address as `clickedValidationAddressOption`
- use proper action types and rename the key from `action_type` to `action`

Summing up the changes to behavior as viewed by analytics:
1. action_type has been replaced with the correct key action
2. action_type: clickedValidationAddress is now the corrected action: clickedValidationAddressOptions
3. Closing modal via X behavior should be the same
4. Closing modal via 'back to edit' was previously action_type: clickedValidationAddress  (see 2 above) - now it will be action: clickedCloseValidationAddress with label 'Back to Edit'

This is ready for review and merge pending the comments i've left below.

File: src/Apps/Order/Components/AddressVerificationFlow.tsx (hunk 10 above)
Comment (erikdstock) [reply src/Apps/Order/Components/AddressVerificationFlow.tsx:265]: Confirmed behavior with dayane and updated per her request!

File: src/Apps/Order/Components/AddressVerificationFlow.tsx (hunk 9 above)
Comment (erikdstock) [reply src/Apps/Order/Components/AddressVerificationFlow.tsx]: resolved in latest commit [9bb308f](https://github.com/artsy/force/pull/12778/commits/9bb308f0717dc996fc0314135e65404ad56d3a85) after working with dayane.

File: src/Apps/Order/Components/AddressVerificationFlow.tsx (hunk 8 above)
Comment (erikdstock) [reply src/Apps/Order/Components/AddressVerificationFlow.tsx]: Correct, fixed with a type coercion for now, will ticket updating cohesion to permit this.

File: src/Apps/Order/Components/AddressVerificationFlow.tsx (hunk 6 above)
Comment (erikdstock) [reply src/Apps/Order/Components/AddressVerificationFlow.tsx]: Updated to simply fire once on mount with `[]`.

File: src/Apps/Order/Components/AddressVerificationFlow.tsx (hunk 14 above)
Comment (erikdstock) [reply src/Apps/Order/Components/AddressVerificationFlow.tsx:166]: If we come back to this and want to add values that are not present immediately on render (eg tracking a value that is null to start but then changed with a `set_` we would need the additional `useRef` that I removed in that commit 👆 . Otherwise this is the right way to do it according to web-practice.

Comment (starsirius) [reply src/Apps/Order/Components/AddressVerificationFlow.tsx:166]: Yeah I wasn't sure if there will be cases that modal is not actually rendered, or more impression tracking fired than necessary when react batches state update. Sounds like it's the recommended way. 👍 

Linking to the [discussion](https://artsy.slack.com/archives/C2TQ4PT8R/p1693247057861649) for completeness.
